From e508559113c2b551fbc73ee48bb8ba6d675f7c49 Mon Sep 17 00:00:00 2001
From: Phantom X <PhantomX@users.noreply.github.com>
Date: Tue, 21 Jul 2020 11:18:03 -0300
Subject: [PATCH] libchdr: system libraries

Use system libchdr or system flac and zlib, if using internal
---
 yabause/src/CMakeLists.txt               | 37 ++++++++++++++++++++----
 yabause/src/tools/libchdr/CMakeLists.txt | 23 +++++++++++++--
 2 files changed, 51 insertions(+), 9 deletions(-)

diff --git a/yabause/src/CMakeLists.txt b/yabause/src/CMakeLists.txt
index ae57b3e..7d0d83a 100755
--- a/yabause/src/CMakeLists.txt
+++ b/yabause/src/CMakeLists.txt
@@ -1,5 +1,9 @@
 project(kronos)
 
+find_package(PkgConfig REQUIRED)
+OPTION(USE_SYSTEM_CHDR "Dynamically link against system libchdr" ON)
+OPTION(USE_SYSTEM_FLAC "Dynamically link against system flac" ON)
+
 #to compile sdl port, use "cmake -DYAB_PORTS=sdl -DYAB_ASYNC_RENDERING ../kronos"
 
 include (CheckCSourceCompiles)
@@ -140,12 +144,33 @@ set(kronos_SOURCES
 	)
 add_definitions(-DIMPROVED_SAVESTATES)
 
-add_subdirectory(tools/libchdr)
-include_directories(${CMAKE_CURRENT_SOURCE_DIR}/tools/libchdr/src)
-set(KRONOS_LIBRARIES ${KRONOS_LIBRARIES} chdr-static)
-
-include_directories(${CMAKE_CURRENT_SOURCE_DIR}/tools/libchdr/deps/zlib-1.2.11)
-set(KRONOS_LIBRARIES ${KRONOS_LIBRARIES} zlibstatic)
+if (USE_SYSTEM_CHDR)
+  message(STATUS "Using shared libchdr")
+  add_library(chdr-shared INTERFACE IMPORTED GLOBAL)
+  pkg_check_modules(CHDR REQUIRED libchdr)
+  include_directories(${CHDR_INCLUDE_DIRS})
+  set(KRONOS_LIBRARIES ${KRONOS_LIBRARIES} ${CHDR_LIBRARIES})
+
+else(USE_SYSTEM_CHDR)
+
+  add_subdirectory(tools/libchdr)
+  include_directories(${CMAKE_CURRENT_SOURCE_DIR}/tools/libchdr/src)
+  set(KRONOS_LIBRARIES ${KRONOS_LIBRARIES} chdr-static)
+  if(USE_SYSTEM_FLAC)
+    message(STATUS "Using shared flac")
+    set(KRONOS_LIBRARIES ${KRONOS_LIBRARIES} flac-static)
+  endif()
+
+endif(USE_SYSTEM_CHDR)
+
+find_package(ZLIB)
+if(ZLIB_FOUND)
+  message(STATUS "Using shared zlib")
+  set(KRONOS_LIBRARIES ${KRONOS_LIBRARIES} ZLIB::ZLIB)
+else()
+  include_directories(${CMAKE_CURRENT_SOURCE_DIR}/tools/libchdr/deps/zlib-1.2.11)
+  set(KRONOS_LIBRARIES ${KRONOS_LIBRARIES} zlibstatic)
+endif()
 
 include_directories(tools/nanovg)
 set(kronos_SOURCES ${kronos_SOURCES} tools/nanovg/nanovg.c)
diff --git a/yabause/src/tools/libchdr/CMakeLists.txt b/yabause/src/tools/libchdr/CMakeLists.txt
index 34df19c..c33e07c 100644
--- a/yabause/src/tools/libchdr/CMakeLists.txt
+++ b/yabause/src/tools/libchdr/CMakeLists.txt
@@ -1,5 +1,7 @@
 cmake_minimum_required(VERSION 3.1)
 
+find_package(PkgConfig REQUIRED)
+
 project(chdr C)
 
 #--------------------------------------------------
@@ -16,6 +18,7 @@ list(APPEND CHDR_INCLUDES deps/crypto)
 list(APPEND CHDR_LIBS crypto-static)
 
 # flac
+if(NOT USE_SYSTEM_FLAC)
 set(FLAC_SOURCES
   deps/flac-1.3.2/src/libFLAC/bitmath.c
   deps/flac-1.3.2/src/libFLAC/bitreader.c
@@ -61,6 +64,13 @@ target_compile_definitions(flac-static PRIVATE PACKAGE_VERSION="1.3.2" FLAC__HAS
 list(APPEND CHDR_INCLUDES deps/flac-1.3.2/src/libFLAC/include)
 list(APPEND CHDR_DEFS FLAC__NO_DLL)
 list(APPEND CHDR_LIBS flac-static)
+else()
+  message(STATUS "libchdr - Using shared flac")
+  add_library(flac-static INTERFACE IMPORTED GLOBAL)
+  pkg_check_modules(FLAC REQUIRED flac)
+  list(APPEND CHDR_INCLUDES ${FLAC_INCLUDE_DIRS})
+  list(APPEND CHDR_LIBS ${FLAC_LIBRARIES})
+endif()
 
 # lzma
 set(LZMA_SOURCES
@@ -114,9 +124,16 @@ list(APPEND CHDR_INCLUDES deps/lzma-16.04/C)
 list(APPEND CHDR_LIBS lzma-static)
 
 # zlib
-add_subdirectory(deps/zlib-1.2.11 EXCLUDE_FROM_ALL)
-list(APPEND CHDR_INCLUDES deps/zlib-1.2.11 ${CMAKE_CURRENT_BINARY_DIR}/deps/zlib-1.2.11)
-list(APPEND CHDR_LIBS zlibstatic)
+find_package(ZLIB)
+if(ZLIB_FOUND)
+  message(STATUS "libchdr - using shared zlib")
+  list(APPEND CHDR_LIBS ZLIB::ZLIB)
+else()
+  add_subdirectory(deps/zlib-1.2.11 EXCLUDE_FROM_ALL)
+  list(APPEND CHDR_INCLUDES deps/zlib-1.2.11 ${CMAKE_CURRENT_BINARY_DIR}/deps/zlib-1.2.11)
+  list(APPEND CHDR_LIBS zlibstatic)
+endif()
+
 
 #--------------------------------------------------
 # chdr
-- 
2.26.2

